src/trackinfoobject.cpp

```diff
@@ -110,26 +101,49 @@ void TrackInfoObject::setDeleteOnReferenceExpiration(bool deleteOnReferenceExpir
     m_bDeleteOnReferenceExpiration = deleteOnReferenceExpiration;
 }
 
-void TrackInfoObject::setMetadata(const Mixxx::TrackMetadata& trackMetadata) {
-    // TODO(XXX): This involves locking the mutex for every setXXX
-    // method. We should figure out an optimization where there are private
-    // setters that don't lock the mutex.
-    setArtist(trackMetadata.getArtist());
-    setTitle(trackMetadata.getTitle());
-    setAlbum(trackMetadata.getAlbum());
-    setAlbumArtist(trackMetadata.getAlbumArtist());
-    setYear(trackMetadata.getYear());
-    setGenre(trackMetadata.getGenre());
-    setComposer(trackMetadata.getComposer());
-    setGrouping(trackMetadata.getGrouping());
-    setComment(trackMetadata.getComment());
-    setTrackNumber(trackMetadata.getTrackNumber());
-    setTrackTotal(trackMetadata.getTrackTotal());
-    setChannels(trackMetadata.getChannels());
-    setSampleRate(trackMetadata.getSampleRate());
-    setDuration(trackMetadata.getDuration());
-    setBitrate(trackMetadata.getBitrate());
-    setReplayGain(trackMetadata.getReplayGain());
+void TrackInfoObject::setTrackMetadata(
+        const Mixxx::TrackMetadata& trackMetadata,
+        QImage *pCoverArt,
+        bool parsedFromFile) {
+    {
+        // enter locking scope
+        QMutexLocker lock(&m_qMutex);
+
+        bool modified = compareAndSet(&m_bHeaderParsed, parsedFromFile);
+        bool modifiedReplayGain =
+                (m_metadata.getReplayGain() != trackMetadata.getReplayGain());
+        bool modifiedCoverArt = false;
+        if (m_metadata != trackMetadata) {
+            const Mixxx::ReplayGain replayGainBackup(m_metadata.getReplayGain());
+            m_metadata = trackMetadata;
+            if (!trackMetadata.getReplayGain().hasRatio()) {
+                // restore replay gain
+                m_metadata.setReplayGain(replayGainBackup);
+                modifiedReplayGain = false;
+            }
+            modified = true;
+        }
+        if (pCoverArt && !pCoverArt->isNull()) {
+            m_coverArt.image = *pCoverArt;
+            m_coverArt.info.hash = CoverArtUtils::calculateHash(
+                m_coverArt.image);
+            m_coverArt.info.coverLocation = QString();
+            m_coverArt.info.type = CoverInfo::METADATA;
+            m_coverArt.info.source = CoverInfo::GUESSED;
```

Comment:
This works here, because setTrackMetadata is only called from SoundSourceProxy::loadTrackMetadataAndCoverArt, and relies on pCoverArt being null when we have already an alternative cover set. 
I think this should moved probably to readTrackMetadataAndCoverArtFromFile()